app/models/workers_scaling/algorithm_runner.rb

```diff
@@ -0,0 +1,77 @@
+##
+# Scheduler to run WorkersScalingAlgorithm decision loop periodically
+module WorkersScaling
+  class AlgorithmRunner
+
+    ##
+    # Returns instance of AlgorithmRunner for experiment_id if registered
+    # Otherwise returns nil
+    def self.get(experiment_id)
+      return @@cache[experiment_id] if @@cache
+    end
+
+    ##
+    # Registers instance of AlgorithmRunner for experiment_id
+    def self.put(experiment_id, runner)
+      @@cache ||= {}
+      @@cache[experiment_id] = runner
+    end
+
+    ##
+    # Unregisters instance of AlgorithmRunner for experiment_id
+    def self.delete(experiment_id)
+      @@cache.delete(experiment_id) if @@cache
+    end
+
+    ##
+    # Expects:
+    # * instance of Experiment
+    # * instance of WorkersScalingAlgorithm for given Experiment
+    # * interval between decision loop executions in seconds
+    def initialize(experiment, algorithm, interval)
+      @experiment = experiment
+      @algorithm = algorithm
+      @interval = interval
+      @mutex = Mutex.new
+
+      self.class.put(@experiment.id, self)
+    end
+
+    ##
+    # Initializes WorkersScalingAlgorithm, then starts its decision loop
+    # Stops when there is no next execution time set
+    def start
+      Thread.new do
+        @algorithm.initial_deployment
+        @next_execution_time = Time.now + @interval
+
+        until @next_execution_time.nil?
+
+          while @next_execution_time > Time.now do
```

Comment:
Racja. Chociaż jedyne, co może się tutaj wydarzyć, to odczytanie zmiennej nieaktualnej o ułamek sekundy. Mimo to dodam mutex.